File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_cpp/code_output/KappaCalculator.cpp
#include <iostream>
#include <vector>
#include <numeric>

double kappa(const std::vector<std::vector<int>>& testData, int k) {
    const auto& dataMat = testData;
    double P0 = 0.0;
    for (int i = 0; i < k; i++) {
        P0 += dataMat[i][i] * 1.0;
    }
    std::vector<int> xsum(k, 0);
    std::vector<int> ysum(k, 0);
    int sum = 0;
    for (int i = 0; i < k; i++) {
        for (int j = 0; j < k; j++) {
            xsum[i] += dataMat[i][j];
            ysum[j] += dataMat[i][j];
            sum += dataMat[i][j];
        }
    }
    double Pe = 0.0;
    for (int i = 0; i < k; i++) {
        Pe += static_cast<double>(ysum[i]) * xsum[i];
    }
    Pe = Pe / (static_cast<double>(sum) * sum);
    P0 = P0 / sum;
    return (P0 - Pe) / (1.0 - Pe);
}

double fleissKappa(const std::vector<std::vector<int>>& testData, int N, int k, int n) {
    const auto& dataMat = testData;
    std::vector<double> P(N, 0.0);
    double sum = 0.0;
    for (int i = 0; i < N; i++) {
        double temp = 0.0;
        for (int j = 0; j < k; j++) {
            sum += dataMat[i][j];
            temp += static_cast<double>(dataMat[i][j]) * dataMat[i][j];
        }
        temp -= n;
        temp /= static_cast<double>((n - 1) * n);
        P[i] = temp;
    }
    double P0 = std::accumulate(P.begin(), P.end(), 0.0) / N;
    std::vector<double> pj(k, 0.0);
    for (int j = 0; j < k; j++) {
        for (int i = 0; i < N; i++) {
            pj[j] += dataMat[i][j];
        }
        pj[j] /= sum;
    }
    double Pe = 0.0;
    for (int j = 0; j < k; j++) {
        Pe += pj[j] * pj[j];
    }
    return (P0 - Pe) / (1.0 - Pe);
}

int main() {
    std::cout << kappa({{2, 1, 1}, {1, 2, 1}, {1, 1, 2}}, 3) << std::endl; // 0.25
    std::cout << fleissKappa({{0, 0, 0, 0, 14},
                              {0, 2, 6, 4, 2},
                              {0, 0, 3, 5, 6},
                              {0, 3, 9, 2, 0},
                              {2, 2, 8, 1, 1},
                              {7, 7, 0, 0, 0},
                              {3, 2, 6, 3, 0},
                              {2, 5, 3, 2, 2},
                              {6, 5, 2, 1, 0},
                              {0, 2, 2, 3, 7}}, 10, 5, 14) << std::endl;
    return 0;
}

File: data/derived_cot/rq1_traces/translation/glm5.1/java_to_py/code_output/KappaCalculator.py
def kappa(test_data, k):
    data_mat = test_data
    p0 = 0.0
    for i in range(k):
        p0 += data_mat[i][i]

    xsum = [0] * k
    ysum = [0] * k
    total_sum = 0
    for i in range(k):
        for j in range(k):
            xsum[i] += data_mat[i][j]
            ysum[j] += data_mat[i][j]
            total_sum += data_mat[i][j]

    pe = 0.0
    for i in range(k):
        pe += ysum[i] * xsum[i]

    pe = pe / (total_sum * total_sum)
    p0 = p0 / total_sum
    return (p0 - pe) / (1 - pe)


def fleiss_kappa(test_data, N, k, n):
    data_mat = test_data
    P = [0.0] * N
    total_sum = 0.0
    for i in range(N):
        temp = 0.0
        for j in range(k):
            total_sum += data_mat[i][j]
            temp += data_mat[i][j] * data_mat[i][j]
        temp -= n
        temp /= (n - 1) * n
        P[i] = temp

    p0 = sum(P) / N

    pj = [0.0] * k
    for j in range(k):
        for i in range(N):
            pj[j] += data_mat[i][j]
        pj[j] /= total_sum

    pe = sum(p * p for p in pj)
    return (p0 - pe) / (1 - pe)


if __name__ == "__main__":
    print(kappa([[2, 1, 1], [1, 2, 1], [1, 1, 2]], 3))  # 0.25
    print(fleiss_kappa([[0, 0, 0, 0, 14],
                        [0, 2, 6, 4, 2],
                        [0, 0, 3, 5, 6],
                        [0, 3, 9, 2, 0],
                        [2, 2, 8, 1, 1],
                        [7, 7, 0, 0, 0],
                        [3, 2, 6, 3, 0],
                        [2, 5, 3, 2, 2],
                        [6, 5, 2, 1, 0],
                        [0, 2, 2, 3, 7]], 10, 5, 14))